Ready: TS List Keys + Lite/Immutable TS data classes.

alexmoore opened the pull request.

srgg committed:
Introduces XLight TimeSeries API

alexmoore committed:
Boilerplate for TS List Keys

alexmoore committed:
Renaming a few classes, might have more than 1 immutable type in the future

alexmoore committed:
Use IQueryResult for ListKeys result, make ImmutablePbResultFactory / QueryResult more flexible for input

alexmoore committed:
Interface all the TS things!

srgg commented on src/main/java/com/basho/riak/client/core/converters/TimeSeriesPBConverter.java:
```diff
@@ -21,28 +18,28 @@
 {
     private TimeSeriesPBConverter() {}
```
IMO better to introduce complete Abstract Factory pattern in case we decide to move forward with both mutable and immutable APIs

alexmoore replied:
Yeah, that's the way that class seems to be headed.

srgg commented on src/main/java/com/basho/riak/client/core/operations/ts/ListKeysOperation.java:
```diff
@@ -0,0 +1,97 @@
+    private final Builder builder;
+    private BinaryValue queryInfoMessage;
+
+    ListKeysOperation(Builder builder)
```
Suspicious package visibility, do we need it?

alexmoore replied:
Eh, some operations are package local, some are private. Should probably be private though.

srgg commented on src/main/java/com/basho/riak/client/core/query/timeseries/immutable/pb/QueryResult.java:
```diff
@@ -0,0 +1,112 @@
+    {
+        this.pbColumnDescriptions = columnsList;
+        this.pbRows = rowsList;
+        this.rows = new ArrayList<IRow>(this.pbRows.size());
```
From the performance perspective  better do not perform premature initialization. Initialization should be performed on demand in public List<IRow> getRows()

srgg commented on src/main/java/com/basho/riak/client/core/query/timeseries/immutable/pb/QueryResult.java:
```diff
@@ -0,0 +1,112 @@
+        return this.rows;
+    }
+
+    private static class ImmutableRowIterator implements Iterator<IRow>
```
It will be nice to put nested classes at the same place either on top (preferred by me) or at the bottom of the class. 

Row. ImmutableCellIterator vs this one

alexmoore committed:
Take out the interfaces, make the "Light" classes the new default TS classes.

alexmoore committed:
Rename & split ImmutablePbResultFactory, reformat & cleanup.

Rename & split ImmutablePbResultFactory into PbResultFactory and
CollectionConverters.

IntelliJ Reformat, & cleanup.

alexmoore committed:
Fix some test failures around Null cells

alexmoore committed:
Make ListKeysOp ctor private

srgg committed:
Reduce memory consumption in from/to PB Cell/Row conversions

hmmr commented:
Per https://github.com/basho/riak_kv/pull/1291, there will be no non-streaming version of list_keys for TS. Apologies if you guys are already aware and dealing with this.

alexmoore commented:
@srgg Looks like a convertible iterator won't work, the protobuf library iterates over the collection twice: https://github.com/google/protobuf/blob/v2.5.0/java/src/main/java/com/google/protobuf/AbstractMessageLite.java#L323 https://github.com/google/protobuf/blob/v2.5.0/java/src/main/java/com/google/protobuf/AbstractMessageLite.java#L329

I have some time tonight, I'll see if I can switch it to a convertable iterable so we can get it in.

srgg commented:
@alexmoore Damn... will also take a look at it

alexmoore committed:
Introducing wrapped iterables, since PB looks at the iterator at least twice.

alexmoore committed:
Add list keys tests

srgg committed:
Code cleanup, minor renaming, getting rid of cached deep copies.

alexmoore committed:
Fix test

srgg committed:
Getting rid of BinaryValue in TS API.

alexmoore committed:
Merge branch 'ts-list-keys' of https://github.com/basho/riak-java-client into ts-list-keys

srgg commented on src/main/java/com/basho/riak/client/core/query/timeseries/PbResultFactory.java:
```diff
@@ -0,0 +1,52 @@
+            size += resp.getKeysCount();
+        }
+
+        final ArrayList<RiakTsPB.TsRow> pbRows = new ArrayList<RiakTsPB.TsRow>(size);
```
Shall we refactor this logic to Convertible one?

alexmoore replied:
So a ConvertableIteratorIterator<T> ?

srgg replied:
Yep

alexmoore replied:
Sounds like fun, we should look into Guava in the future.

alexmoore replied:
@srgg Actually I think it's better to copy here for now. If we do a ConvertableIteratorIterator we'll still have to copy it to a list for the QueryResult constructor, or completely rewrite QueryResult to use an Iterator<PBRow> under the covers. Thoughts?

srgg commented on src/test/java/com/basho/riak/client/core/operations/itest/ts/ITestListKeysOperation.java:
```diff
@@ -0,0 +1,72 @@
+    @BeforeClass
+    public static void InsertData() throws ExecutionException, InterruptedException
+    {
+        StoreOperation storeOp = new StoreOperation.Builder(tableName).withRows(rows).build();
```
ITestBase.executeQuery() might be used

srgg commented on src/test/java/com/basho/riak/client/core/operations/itest/ts/ITestListKeysOperation.java:
```diff
@@ -0,0 +1,72 @@
+    @Test
+    public void testSingleFetch() throws ExecutionException, InterruptedException
+    {
+        ListKeysOperation listKeysOp = new ListKeysOperation.Builder(tableName).build();
```
ITestBase.executeQuery() might be used

srgg commented on src/test/java/com/basho/riak/client/core/operations/itest/ts/ITestListKeysOperation.java:
```diff
@@ -0,0 +1,72 @@
+        assertTrue(rows.containsAll(expectedKeys));
+    }
+
+    private List<Row> getKeyHeads()
```
Better to make it static

srgg commented:
+1 with that change https://github.com/basho/riak-java-client/pull/579#discussion_r46983904

alexmoore commented:
I can't wait until we're JDK8+ and we can use the streaming apis.

alexmoore committed:
Implement SelectMany in Java

alexmoore committed:
Adjust tests for Eventual Consistency of ListKeys

srgg commented:
May be it makes sense to move anything related to the Magic Iterable to the specific packge

christophermancini commented:
:+1: Tests pass and looks like your tests provide good coverage of the effected code.

alexmoore commented:
Thanks for the review @christophermancini & @srgg.
